exec-service-wasmer/src/wasmer_opcode_control.rs

```diff
@@ -0,0 +1,210 @@
+            OPCODE_CONTROL_MEMORY_GROW.to_string(),
+            ExportIndex::Global(memory_grow_global_index),
+        );
+
+        let memory_grow_cached_global_index = module_info
+            .globals
+            .push(GlobalType::new(Type::I64, Mutability::Var));
+
+        module_info
+            .global_initializers
+            .push(GlobalInit::I64Const(0));
+
+        module_info.exports.insert(
+            OPCODE_CONTROL_MEMORY_GROW_CACHED.to_string(),
+            ExportIndex::Global(memory_grow_cached_global_index),
+        );
+
+        *global_indexes = Some(OpcodeControlGlobalIndexes(
+            memory_grow_global_index,
+            memory_grow_cached_global_index,
+        ));
+    }
+}
+
+impl MemoryUsage for OpcodeControl {
+    fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
+        mem::size_of_val(self) + self.global_indexes.size_of_val(tracker)
+            - mem::size_of_val(&self.global_indexes)
+    }
+}
+
+#[derive(Debug)]
+struct FunctionOpcodeControl {
+    max_memory_grow: usize,
+    max_memory_grow_delta: usize,
+    breakpoint: Arc<Breakpoint>,
+    global_indexes: OpcodeControlGlobalIndexes,
+}
+
+impl FunctionMiddleware for FunctionOpcodeControl {
+    fn feed<'b>(
+        &mut self,
+        operator: Operator<'b>,
+        state: &mut MiddlewareReaderState<'b>,
+    ) -> Result<(), MiddlewareError> {
+        match operator {
+            Operator::MemoryGrow { .. } => {
+                // Check if memory limit (memory_grow >= max_memory_grow)
+                state.extend(&[
+                    Operator::GlobalGet {
+                        global_index: self.global_indexes.memory_grow().as_u32(),
+                    },
+                    Operator::I64Const {
+                        value: self.max_memory_grow as i64,
+                    },
+                    Operator::I64GeU,
+                ]);
+
+                // Insert breakpoint BREAKPOINT_VALUE_MEMORY_LIMIT if memory_grow >= max_memory_grow
+                state.extend(
+                    self.breakpoint
+                        .as_ref()
+                        .insert_breakpoint(BREAKPOINT_VALUE_MEMORY_LIMIT)
+                        .iter(),
+                );
+
+                // Increment the memory_grow counter
+                state.extend(&[
+                    Operator::GlobalGet {
+                        global_index: self.global_indexes.memory_grow().as_u32(),
+                    },
+                    Operator::I64Const { value: 1 as i64 },
+                    Operator::I64Add,
+                    Operator::GlobalSet {
+                        global_index: self.global_indexes.memory_grow().as_u32(),
+                    },
+                ]);
+
+                // Cache the memory_grow counter
+                state.extend(&[Operator::GlobalSet {
+                    global_index: self.global_indexes.memory_grow_cached().as_u32(),
+                }]);
```

Comment:
Comment warning: this is not the `memory_grow` counter (the one compared to the memory limit above).

It is the `memory_grow_delta`, namely the argument received by `memory_grow`. It's the `delta` in the `memory.grow delta` WASM opcode.

Reply:
Oh, I get it now. Thanks for pointing out. Will refactor into descriptive methods anyway.